### PR TITLE
Workshop 2

### PR DESCRIPTION
**Pull request by Smaktulka**
Add custom exception (DtoScanException), functionality for saving and loading repositories.

### CODE REVIEW

File: workshop_1/src/main/java/by/andersen/context/AppContext.java [hunk 1]
```diff
@@ -24,6 +30,33 @@ public static AppContext init() {
...
+      RepositoryContext repositoryContext,
+      Path repositoryStateFilePath
+  ) {
+    if (Files.exists(repositoryStateFilePath)) {
+      repositoryContext.loadFromFile(repositoryStateFilePath);
+    } else {
+      try {
+        Files.createFile(repositoryStateFilePath);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```
Comment (elnurmes) [inline workshop_1/src/main/java/by/andersen/context/AppContext.java]: avoid generic RuntimeException, at least use new IllegalStateException with some message

File: workshop_1/src/main/java/by/andersen/context/RepositoryContext.java [hunk 2]
```diff
@@ -1,17 +1,44 @@
...
+
+  @SuppressWarnings("unchecked")
+  public void loadFromFile(Path filePath) {
+    try (ObjectInputStream ois =
+        new ObjectInputStream(Files.newInputStream(filePath))) {
+      Map<Class<?>, Repository<?, ?>> loadedRepositories = (Map<Class<?>, Repository<?, ?>>) ois.readObject();
+      repositoryMap.clear();
+      repositoryMap.putAll(loadedRepositories);
+    } catch (IOException | ClassNotFoundException e) {
+      throw new RuntimeException(e);
```
Comment (elnurmes) [inline workshop_1/src/main/java/by/andersen/context/RepositoryContext.java]: same here

File: workshop_1/src/main/java/by/andersen/utils/PropertiesUtils.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+public class PropertiesUtils {
+  private Properties properties;
+  private static final String APP_PROPERTIES_FILENAME = "app.properties";
+  private static final String REPOSITORY_STATE_FILEPATH = "repository.state.filepath";
+
+  public static PropertiesUtils loadAppPropertiesFile() {
+    Properties properties = new Properties();
+    Path appPropertiesFile = getAppPropertiesFile();
+    try {
+      properties.load(Files.newInputStream(appPropertiesFile));
```
Comment (elnurmes) [inline workshop_1/src/main/java/by/andersen/utils/PropertiesUtils.java]: use try-with resources here

File: workshop_1/src/main/java/by/andersen/utils/PropertiesUtils.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    return Optional.of(repositoryStateFilePath);
+  }
+
+  private static Path getAppPropertiesFile() {
+    URI appPropertiesUri;
+    try {
+      appPropertiesUri = ClassLoader.getSystemResource(APP_PROPERTIES_FILENAME).toURI();
+    } catch (URISyntaxException e) {
+      throw new RuntimeException(e);
```
Comment (elnurmes) [inline workshop_1/src/main/java/by/andersen/utils/PropertiesUtils.java]: same about exception

**Review by elnurmes — CHANGES_REQUESTED** [review]
hi, I added a few small comments, pls check it

**Comment by Smaktulka** [comment]
Can you please explain, why it is more preferrable to use IllegalStateException? It should increase readability?

**Comment by elnurmes** [comment]
> Can you please explain, why it is more preferrable to use IllegalStateException? It should increase readability?
You are working with app states and its provides clearer context for this purpose. So in this case IllegalStateException signaling invalid states, which is more related to your context. Additionally RuntimeException is more generic, so using it could increase problems with debugging

**Review by elnurmes — APPROVED** [review]
Hi, you can merge it, great job!

**Pull request merged**